README.md
[![NPM module](https://badge.fury.io/js/bigval.svg)](https://badge.fury.io/js/bigval)
[![Join the community](https://img.shields.io/badge/Chat%20on-Telegram-brightgreen.svg?color=0088cc)](https://t.me/erdDEV)
[![Follow on Twitter](https://img.shields.io/twitter/url/http/shields.io.svg?style=social&label=Follow&maxAge=2592000)](https://twitter.com/erd_dev)

# bigval

Javascript library for representing arbitrarily large or small crypto values with decimals. 

`bigval` is useful for handling extremely large crypto numbers with precision, e.g. when performing arithmetic with balances.

It is inspired by [ethval](https://github.com/hiddentao/ethval).

Features:

* Built-on on the robust [decimal.js](https://mikemcl.github.io/decimal.js/) library.
* Immutable-by-default to maximize code safety.
* Flexible about input types.
* Easy conversion between number scales (see below).

## Installation

```
npm install --save bigval
```

## Usage

Initializing a value:

```js
const { BigVal } = require('bigval')

const v1 = new BigVal(1000000000000000000)
const v2 = new BigVal('1000000000000000000')
const v3 = new BigVal('0.00000001')
const v4 = new BigVal('0xde0b6b3a7640000')
const v5 = new BigVal('b1000000000000000000')
const v6 = new BigVal(new BigVal(1000000000000000000))
```


Scaling:

```js
const v6 = new BigVal(6, 'coins')
const sameAsV6 = new BigVal('6000000000000000000', 'min')
const sameAsV6Again = new BigVal('6000000000000000000')
```

Scaling is based on no. of decimals:

```js
const v = new BigVal(100, 'min', { decimals: 2 }) // same as new BigVal(1, 'coins', { decimals: 2 })

console.log( v.scale ) // 'min'

const v2 = v.toCoinScale()

console.log( v2.scale ) // 'coins'
console.log( v2.toString() ) // "1"

const v3 = v2.toMinScale()

console.log( v3.scale ) // 'min'
console.log( v3.toString() ) // "100"
```

Simple arithmetic leaves original unchanged:

```js
const v = new BigVal(1000000000000000000)

const v2 = v.add(32) // v2 is a new instance of BigVal

console.log(v2 !== v) // true

console.log(v2.toString()) // "1000000000000000032"

console.log(v.toString()) // "1000000000000000000" - original unchanged
```

Initializing using `from()` static convenience method:

```js
const v6 = BigVal::from('1', 'coins', { decimals: 2 })
const v6_1 = v6.toMinScale()

console.log( v6_1.scale ) // 'min'
console.log( v6_1.toString() ) // "100"
```

Initializing using `fromStr()` static method:

```js
const v7 = BigVal::fromStr('2 coins') // same as BigVal::from(2, 'coins')
const v8 = BigVal::fromStr('0.5 min') // same as BigVal::from(0.5, 'min')
const v9 = BigVal::fromStr('0xff', { decimals: 2 }) // same as BigVal::from(0xff, undefined, { decimals: 2 })
```

Output in different types:

```js
const v = new BigVal(255)

console.log( v.toString() ) // "255"
console.log( v.toString(10) ) // "255"
console.log( v.toString(16) ) // "FF"
console.log( v.toString(2) ) // "1111111"
console.log( v.toNumber() ) // 255
```

Decimals and rounding:

```js
const v = new BigVal(100)

const v2 = v.div(3)

console.log( v2.toString() ) // "33.333333333333333"
console.log( v2.toFixed(1) ) // "33.3"

const v3 = v2.round()

console.log( v3.toString() ) // "33"

const v4 = new BigVal(0.00000001)
console.log(v4.decimalCount) // 8
```


Flexible input types:

```js
const v = new BigVal(255)

// these are all the same
v.add(32)
v.add('32')
v.add(new BigVal(32))
```

Logical operators:

```js
const v = new BigVal(255)

const isGreater = v.gte(254)

console.log( isGreater ) // true
```

## Number scales

At any given time a `BigVal` instance operates at a particular number _scale_. The scale is based on the the no. of `decimals` specified in the configuration (`BigValConfig`). By default the no. of `decimals` is 18.
 
The `min` scale (this is the default) is for numbers which do not have decimal places since they are already denominated in the smallest possible unit. The `coins` scale is for numbers which implicitly have decimal places.
 
For example, if a given `BigVal` has `decimals = 2` then the following two numbers are equivalent in value:
 
 * scale = `min`, value = `100`
 * scale = `coins`, value = `1`

If `decimals = 18` (this is the default) then the following two numbers are equivalent in value:

* scale = `min`, value = `1000000000000000000`
* scale = `coins`, value = `1`
 
The use of scales like this makes it easy to convert between chain-friendly and user-friendly values and perform arithmetic at the desired precision.


## Developer guide

To build both ESM and CommonJS output:

```shell
yarn build
```

To re-build the CommonJS output on chnage:

```shell
yarn dev
```

To test:

```shell
yarn test
```

To build the docs:

```shell
yarn build-docs
```

To publish a new release (this will create a tag, publish to NPM and publish the latest docs):

```shell
yarn release
```

## License

MIT

CHANGELOG.md
# [1.7.0](https://github.com/hiddentao/bigval/compare/v1.6.0...v1.7.0) (2023-11-06)

# [1.6.0](https://github.com/hiddentao/bigval/compare/v1.5.0...v1.6.0) (2023-11-06)


### Features

* add decimals getter ([9b3fbb0](https://github.com/hiddentao/bigval/commit/9b3fbb06d0fe9346437a37ced08a92899cc6ec96))

# [1.5.0](https://github.com/hiddentao/bigval/compare/v1.4.0...v1.5.0) (2022-03-02)


### Features

* add static construction methods ([c8bb453](https://github.com/hiddentao/bigval/commit/c8bb45396256200241e37c62ec7219ace1ea1cdb))
* add toMinStr() convenience method ([704e87e](https://github.com/hiddentao/bigval/commit/704e87e1836358dbd3600762e55714e5b0238058))

# [1.4.0](https://github.com/erdDEVcode/bigval/compare/v1.3.1...v1.4.0) (2021-02-04)

## [1.3.1](https://github.com/erdDEVcode/bigval/compare/v1.3.0...v1.3.1) (2021-01-14)


### Bug Fixes

* isBigVal should not panic for empty values ([a49739b](https://github.com/erdDEVcode/bigval/commit/a49739b60a5a36890bd2a611551c591d205e8b65))

# [1.3.0](https://github.com/erdDEVcode/bigval/compare/v1.2.0...v1.3.0) (2021-01-14)


### Features

* isBigVal() check ([70ad610](https://github.com/erdDEVcode/bigval/commit/70ad610b17890d384a2c6d2bdc58b68274e0ef23))

# [1.2.0](https://github.com/erdDEVcode/bigval/compare/v1.1.2...v1.2.0) (2021-01-11)


### Features

* added tests ([60d3f91](https://github.com/erdDEVcode/bigval/commit/60d3f914c343816f85474aab2bbe5cb0c8658cbb))

## [1.1.2](https://github.com/erdDEVcode/bigval/compare/v1.1.1...v1.1.2) (2021-01-09)

## [1.1.1](https://github.com/erdDEVcode/bigval/compare/v1.1.0...v1.1.1) (2021-01-09)


### Bug Fixes

* method name ([824cebd](https://github.com/erdDEVcode/bigval/commit/824cebd9a259a515ef536c7c34ee91aae04ac501))

# [1.1.0](https://github.com/erdDEVcode/bigval/compare/v1.0.2...v1.1.0) (2021-01-09)

## [1.0.2](https://github.com/erdDEVcode/bigval/compare/v1.0.1...v1.0.2) (2021-01-08)


### Bug Fixes

* erroneous reference ([4bd6cc7](https://github.com/erdDEVcode/bigval/commit/4bd6cc793c0d0222481c0db3360990aa6204b97c))

## [1.0.1](https://github.com/erdDEVcode/bigval/compare/v1.0.0...v1.0.1) (2021-01-04)

# 1.0.0 (2021-01-04)


test/index.test.ts
import { BigVal, isBigVal, toMinStr } from '../'

const { expect} = require('./utils')


describe('constructor', () => {
  it('number', () => {
    (new BigVal(1)).should.be.instanceOf(BigVal)
  })

  it('string', () => {
    (new BigVal('1')).should.be.instanceOf(BigVal)
  })

  it('hex', () => {
    (new BigVal('0x1')).should.be.instanceOf(BigVal)
  })

  it('BigVal', () => {
    (new BigVal(new BigVal(1000000000000000))).should.be.instanceOf(BigVal)
  })

  it('Decimal.js', () => {
    (new BigVal((new BigVal(1000000000000000))._n)).should.be.instanceOf(BigVal)
  })
})

describe('output formats', () => {
  it('toNumber()', () => {
    (new BigVal(123)).toNumber().should.equal(123);
    (new BigVal('123')).toNumber().should.equal(123);
    (new BigVal('0x123')).toNumber().should.equal(291);
    (new BigVal('0xFFaab1324')).toNumber().should.equal(68630024996);
    (new BigVal(new BigVal(100000000000000000000))).toNumber().should.equal(100000000000000000000);
    (new BigVal((new BigVal(100000000000000000000))._n)).toNumber().should.equal(100000000000000000000);
  })

  it('toString()', () => {
    (new BigVal(123)).toString().should.equal('123');
    (new BigVal('123')).toString().should.equal('123');
    (new BigVal('0x123')).toString().should.equal('291');
    (new BigVal('0xFFaab1324')).toString().should.equal('68630024996');
    (new BigVal(new BigVal(100000000000000000000))).toString().should.equal('100000000000000000000');
    (new BigVal((new BigVal(100000000000000000000))._n)).toString().should.equal('100000000000000000000');
  })

  it('toString(16)', () => {
    (new BigVal(123)).toString(16).should.equal('0x7b');
    (new BigVal('123')).toString(16).should.equal('0x7b');
    (new BigVal('0x123')).toString(16).should.equal('0x123');
    (new BigVal('0xFFaab1324')).toString(16).should.equal('0xffaab1324');
    (new BigVal(new BigVal(100000000000000000000))).toString(16).should.equal('0x56bc75e2d63100000');
    (new BigVal((new BigVal(100000000000000000000))._n)).toString(16).should.equal('0x56bc75e2d63100000');
  })

  it('toString(2)', () => {
    (new BigVal(123)).toString(2).should.equal('1111011');
    (new BigVal('123')).toString(2).should.equal('1111011');
    (new BigVal('0x123')).toString(2).should.equal('100100011');
    (new BigVal('0xFFaab1324')).toString(2).should.equal('111111111010101010110000000000000000');
    (new BigVal(new BigVal(100000000000000000000))).toString(2).should.equal('1010110101111000111100000000000000000000000000000000000000000000000');
    (new BigVal((new BigVal(100000000000000000000))._n)).toString(2).should.equal('1010110101111000111100000000000000000000000000000000000000000000000');
  })

  it('toFixed()', () => {
    (new BigVal('123.23344')).toFixed(2).should.equal('123.23');
    (new BigVal('123.23584')).toFixed(3).should.equal('123.236');
    (new BigVal('123.23584')).toFixed(0).should.equal('123');
  })
})



describe('scaling + decimals', () => {
  it('construction basics', () => {
    const a = new BigVal("123.4", "coins");
    a.toString().should.equal("123.4");
    a.scale.should.equal("coins");
    a.config.should.deep.equal({ decimals: 18 });
    a.decimalCount.should.equal(1);

    const b = new BigVal("123.4", "min");
    b.toString().should.equal("123.4");
    b.scale.should.equal("min");
    b.config.should.deep.equal({ decimals: 18 });
    b.decimalCount.should.equal(1);

    const c = new BigVal("123.4", "min");
    c.toString().should.equal("123.4");
    c.scale.should.equal("min");
    c.config.should.deep.equal({ decimals: 18 });
    c.decimalCount.should.equal(1);

    const d = new BigVal("123.4", "min", { decimals: 2 });
    d.toString().should.equal("123.4");
    d.scale.should.equal("min");
    d.config.should.deep.equal({ decimals: 2 });
    d.decimalCount.should.equal(1);

    const e = new BigVal("0.00000001", "coins", { decimals: 8 });
    e.toString().should.equal("1e-8");
    e.scale.should.equal("coins");
    e.config.should.deep.equal({ decimals: 8 });
    e.decimalCount.should.equal(8);
  })

  it('constructing from existing BigVal copies source scale', () => {
    const a = new BigVal(1, 'coins')
    const b = new BigVal(a)
    b.scale.should.equal(a.scale)
  })

  describe('static: from()', () => {
    it('construction basics', () => {
      const a = (BigVal.from('123.4', 'coins'));
      a.toString().should.equal('123.4');
      a.scale.should.equal('coins');
      a.config.should.deep.equal({ decimals: 18 });

      const b = (BigVal.from('123.4', 'min'))
      b.toString().should.equal('123.4');
      b.scale.should.equal('min');
      b.config.should.deep.equal({ decimals: 18 });

      const c = (BigVal.from('123.4', 'min'))
      c.toString().should.equal('123.4');
      c.scale.should.equal('min');
      c.config.should.deep.equal({ decimals: 18 });

      const d = (BigVal.from('123.4', 'min', { decimals: 2 }));
      d.toString().should.equal('123.4');
      d.scale.should.equal('min');
      d.config.should.deep.equal({ decimals: 2 });
    })

    it('constructing from existing BigVal copies source scale', () => {
      const a = new BigVal(1, 'coins')
      const b = BigVal.from(a)
      b.scale.should.equal(a.scale)
    })
  })

  describe('static: fromStr()', () => {
    it('construction basics', () => {
      const a = (BigVal.fromStr('123.4 coins'));
      a.toString().should.equal('123.4');
      a.scale.should.equal('coins');
      a.config.should.deep.equal({ decimals: 18 });

      const b = (BigVal.fromStr('123.4 min'))
      b.toString().should.equal('123.4');
      b.scale.should.equal('min');
      b.config.should.deep.equal({ decimals: 18 });

      const c = (BigVal.fromStr('123.4 min'))
      c.toString().should.equal('123.4');
      c.scale.should.equal('min');
      c.config.should.deep.equal({ decimals: 18 });

      const d = (BigVal.fromStr('123.4 min', { decimals: 2 }));
      d.toString().should.equal('123.4');
      d.scale.should.equal('min');
      d.config.should.deep.equal({ decimals: 2 });
    })
  })

  it('scale up/down', () => {
    const a = new BigVal('123.4', 'coins', { decimals: 2 });

    a.toCoinScale().should.not.equal(a) // immutable
    a.toCoinScale().toString().should.equal('123.4')
    a.toCoinScale().scale.should.equal('coins')

    a.toScale('coins').should.not.equal(a) // immutable
    a.toScale('coins').toString().should.equal('123.4')
    a.toScale('coins').scale.should.equal('coins')

    a.toMinScale().should.not.equal(a) // immutable
    a.toMinScale().toString().should.equal('12340')
    a.toMinScale().scale.should.equal('min')

    a.toScale('min').should.not.equal(a) // immutable
    a.toScale('min').toString().should.equal('12340')
    a.toScale('min').scale.should.equal('min')

    a.toMinScale().toCoinScale().toString().should.equal('123.4')
    a.toMinScale().toCoinScale().scale.should.equal('coins')
  })
})


describe('boolean logic', () => {
  it('number', () => {
    const a = new BigVal(100, 'coins')
    a.eq(100).should.be.true
    a.gte(100).should.be.true
    a.gt(100).should.be.false
    a.gt(99).should.be.true
    a.lte(100).should.be.true
    a.lt(100).should.be.false
    a.lt(101).should.be.true
  })

  it('string', () => {
    const a = new BigVal(100, 'coins')
    a.eq('100').should.be.true
    a.gte('100').should.be.true
    a.gt('100').should.be.false
    a.gt('99').should.be.true
    a.lte('100').should.be.true
    a.lt('100').should.be.false
    a.lt('101').should.be.true
  })

  it('hex', () => {
    const a = new BigVal(100, 'coins')
    a.eq('0x64').should.be.true
    a.gte('0x64').should.be.true
    a.gt('0x64').should.be.false
    a.gt('0x63').should.be.true
    a.lte('0x64').should.be.true
    a.lt('0x64').should.be.false
    a.lt('0x65').should.be.true
  })

  it('BigVal', () => {
    const a = new BigVal(100, 'coins')
    a.eq(new BigVal('0x64')).should.be.true
    a.gte(new BigVal('0x64')).should.be.true
    a.gt(new BigVal('0x64')).should.be.false
    a.gt(new BigVal('0x63')).should.be.true
    a.lte(new BigVal('0x64')).should.be.true
    a.lt(new BigVal('0x64')).should.be.false
    a.lt(new BigVal('0x65')).should.be.true
  })
})


describe('arithmetic', () => {
  it('number', () => {
    const a = new BigVal(98)
    const v = 20
    a.add(v).toString().should.equal('118')
    a.sub(v).toString().should.equal('78')
    a.mul(v).toString().should.equal('1960')
    a.div(v).toString(3).should.equal('4.9')
  })

  it('string', () => {
    const a = new BigVal(98)
    const v = '20'
    a.add(v).toString().should.equal('118')
    a.sub(v).toString().should.equal('78')
    a.mul(v).toString().should.equal('1960')
    a.div(v).toString(3).should.equal('4.9')
  })

  it('hex', () => {
    const a = new BigVal(98)
    const v = '0x14'
    a.add(v).toString().should.equal('118')
    a.sub(v).toString().should.equal('78')
    a.mul(v).toString().should.equal('1960')
    a.div(v).toString(3).should.equal('4.9')
  })

  it('BigVal', () => {
    const a = new BigVal(98)
    const v = new BigVal(20)
    a.add(v).toString().should.equal('118')
    a.sub(v).toString().should.equal('78')
    a.mul(v).toString().should.equal('1960')
    a.div(v).toString(3).should.equal('4.9')
  })
})


describe('isBigVal()', () => {
  it('works', () => {
    expect(isBigVal(0)).to.be.false
    expect(isBigVal('100')).to.be.false
    expect(isBigVal({
      toString: () => {}
    })).to.be.false
    expect(isBigVal(undefined)).to.be.false
    expect(isBigVal(null)).to.be.false
    expect(isBigVal(new BigVal(1))).to.be.true
  })
})


describe('toMinStr()', () => {
  it('works', () => {
    expect(toMinStr('0')).to.eq('0')
    expect(toMinStr('100')).to.eq('100')
    expect(toMinStr('100 min')).to.eq('100')
    expect(toMinStr('1 coins')).to.eq('1000000000000000000')
    expect(toMinStr('1 coins', { decimals: 2 })).to.eq('100')
  })
})

src/index.ts
import { Decimal } from 'decimal.js'

/**
 * @internal
 */
const PreciseDecimal = Decimal.clone({ defaults: true, toExpPos: 33 })

/**
 * Convert input to `PreciseDecimal` instance.
 * @param a Input of any type.
 * @internal
 */
const toDecimal = (a: any): Decimal => {
  if (a) {
    if (a._hex) {
      a = a._hex
    } else if (a._isBigNumber) {
      a = a.toString(10)
    }
  }

  return new PreciseDecimal(`${a}`)
}



/**
 * Number scales.
 * @internal
 */
const SCALE: Record<string, string> = {
  coins: 'coins',
  min: 'min',
}

/**
 * Check that given scale is valid.
 * @internal
 */
const assertValidScale = (s: string) => {
  if (!SCALE[s]) {
    throw new Error(`Invalid scale: ${s}`)
  }
}


/**
 * `BigVal` configuration.
 */
export interface BigValConfig {
  /**
   * No. of decimals values will have. Default is 18.
   */
  decimals?: number
}

/**
 * Default config for `BigVal`.
 */
export const DefaultBigValConfig: BigValConfig = { decimals: 18 }


/**
 * Get whether given value is a `BigVal` instance.
 * @param v A value.
 */
export const isBigVal = (v: any): boolean => {
  return !!(v && v._n && v.scale && v.toString && v.toMinScale)
}


/**
 * Represents an arbitrarily large or small number with decimals.
 * 
 * All the arithmetic methods are immutable, i.e. they return a new `BigVal` instance, leaving the original inputs unchanged.
 *
 * At any given time a `BigVal` instance operates at a particular number _scale_. The scale is based on the the no. of `decimals` specified in the configuration (`BigValConfig`).
 * 
 * The `min` scale is for numbers which do not have decimal places since they are already denominated in the smallest possible unit. The `coins` scale is for numbers which implicitly have decimal places.
 * 
 * For example, if a given `BigVal` has `decimals = 2` then the following two numbers are equivalent in value:
 * 
 * * scale = `min`, value = `100`
 * * scale = `coins`, value = `1`
 * 
 * If `decimals = 18` (this is the default) then the following two numbers are equivalent in value:
 * 
 * * scale = `min`, value = `1000000000000000000`
 * * scale = `coins`, value = 1
 * 
 * The use of scales like this makes it easy to convert between chain-friendly and user-friendly values and perform arithmetic at the desired precision.
 */
export class BigVal {
  /**
   * @internal
   */
  _n: Decimal
  /**
   * @internal
   */
  _scale: string
  /**
   * @internal
   */
  _config: BigValConfig
  /**
   * Multiply with another number.
   */
  mul!: (v: any) => BigVal
  /**
   * Subtract another number from this one.
   */
  sub!: (v: any) => BigVal
  /**
   * Divide this by another number.
   */
  div!: (v: any) => BigVal
  /**
   * Add another number to this one.
   */
  add!: (v: any) => BigVal
  /**
   * Get whether this is greater than another number.
   */
  gt!: (v: any) => boolean
  /**
   * Get whether this is greater than or equal to another number.
   */
  gte!: (v: any) => boolean
  /**
   * Get whether this is less than another number.
   */
  lt!: (v: any) => boolean
  /**
   * Get whether this is less than or equal to than another number.
   */
  lte!: (v: any) => boolean
  /**
   * Get whether this is equal to another number.
   */
  eq!: (v: any) => boolean

  /**
   * @constructor
   * @param src Input number. If this is a `BigVal` instance then `scale` and `config` parameters will be ignored.
   * @param scale The scale of the input number. Default is `min`.
   * @param config Custom configuration for this instance.
   */
  constructor(src: any, scale: string = 'min', config: BigValConfig = DefaultBigValConfig) {
    if (isBigVal(src)) {
      this._n = toDecimal(src._n)
      this._scale = src.scale
      this._config = src.config
    } else {
      this._n = toDecimal(src)
      assertValidScale(scale)
      this._scale = scale
      this._config = config
    }

    ;['mul', 'sub', 'div', 'add'].forEach(method => {
      (this as any)[method] = (v: any) => (
        new BigVal((this._n as any)[method].call(this._n, toDecimal(v)), this._scale, this._config)
      )
    })

    ;['gt', 'gte', 'lt', 'lte', 'eq'].forEach(method => {
      (this as any)[method] = (v: any) => (
        (this._n as any)[method].call(this._n, toDecimal(v))
      )
    })
  }

  /**
   * Construct a `BigVal` instance.
   * 
   * @param src Input number. If this is a `BigVal` instance then `scale` and `config` parameters will be ignored.
   * @param scale The scale of the input number. Default is `min`.
   * @param config Custom configuration for this instance.
   * @returns `BigVal` instance.
   */
  static from(src: any, scale: string = 'min', config: BigValConfig = DefaultBigValConfig): BigVal {
    return new BigVal(src, scale, config)
  }

  /**
   * Construct a `BigVal` instance.
   * 
   * @param numScale String in form: `<number> <scale>`. If scale ommitted then `min` is assumed.
   * @param config Custom configuration for this instance.
   * @returns `BigVal` instance.
   */
  static fromStr(numScale: string, config: BigValConfig = DefaultBigValConfig): BigVal {
    const t = numScale.split(' ')
    return this.from(t[0], t[1], config)
  }

  /**
   * Get current scale.
   */
  get scale(): string {
    return this._scale
  }

  /**
   * Get config.
   */
  get config(): BigValConfig {
    return this._config
  }

  /**
   * Get no. of decimal places in the curent value.
   */
  get decimalCount(): number {
    return this._n.decimalPlaces()
  }

  /**
   * Round to the nearest whole number.
   */
  round(): BigVal {
    return new BigVal(this._n.toDecimalPlaces(0), this._scale, this._config)
  }

  /**
   * Convert to 'min' scale.
   */
  toMinScale(): BigVal {
    if (this._scale === SCALE.min) {
      return new BigVal(this)
    } else {
      const n = this.mul(toDecimal(10).pow(toDecimal(this._config.decimals)))
      n._scale = SCALE.min
      return n
    }
  }

  /**
   * Convert to 'coins' scale.
   */
  toCoinScale(): BigVal {
    if (this._scale === SCALE.coins) {
      return new BigVal(this)
    } else {
      const n = this.div(toDecimal(10).pow(toDecimal(this._config.decimals)))
      n._scale = SCALE.coins
      return n
    }
  }

  /**
   * Convert to given scale.
   * 
   * @param scale Scale to convert to.
   */
  toScale(scale: string): BigVal {
    assertValidScale(scale)

    switch (scale) {
      case SCALE.min:
        return this.toMinScale()
      case SCALE.coins:
        return this.toCoinScale()
      default:
        throw new Error(`Unrecognized scale: ${scale}`)
    }
  }

  /**
   * Get string representation in given base.
   * 
   * @param base Base to represent in. Default is 10.
   */
  toString(base: number = 10) {
    switch (base) {
      case 2: {
        let str = this._n.toBinary()
        str = str.substr(str.indexOf('b') + 1)
        return str
      }
      case 16: {
        return this._n.toHexadecimal()
      }
      default:
        return this._n.toString()
    }
  }

  /**
   * Get base-10 string representation to given no. of decimal places.
   * @param numDecimals No. of decimal places to show.
   */
  toFixed(numDecimals: number) {
    return this._n.toFixed(numDecimals)
  }

  /**
   * Get base-10 number representation.
   */
  toNumber() {
    return this._n.toNumber()
  }
}


/**
 * Get string representation of a `BigVal` in `min` scale.
 * 
 * This internally calls `BigVal.fromStr(numScale, config).toMinScale().toString()`.
 * 
 * @param numScale String in form: `<number> <scale>`. If scale ommitted then `min` is assumed.
 * @param config Custom configuration for this instance.
 * @returns string
 */
export const toMinStr = (numScale: string, config: BigValConfig = DefaultBigValConfig): string => {
  return BigVal.fromStr(numScale, config).toMinScale().toString()
}
